README.MD

![Empires & Allies](empiresservercontainer/rte_staging_assets/templates/layouts/logo.png "Empires & Allies logo")
[![Gitpod Ready-to-Code](https://img.shields.io/badge/Gitpod-Ready--to--Code-blue?logo=gitpod)](https://gitpod.io/#https://github.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native)


> Empires and Allies(Facebook) officially closed June 17th, 2013


:world_map: [Main Repo](https://github.com/AcidCaos/raisetheempires)

The Old Empires & Allies preservation project by the Raise the Empires team.
This project is dedicated to preserving and renewing this Flash game so that it's not lost to time.

## Current Status
| Things | Parameter |
| ------- | ------ |
| Current Developed Build | ![version](https://img.shields.io/badge/build-470-blue) |
| Current Stable Releases | ![version](https://img.shields.io/badge/build-470-blue) |
| Current Built Server Size | [![size](https://img.shields.io/badge/size-498%20MB-brightgreen)](#) |
| Current Source Size | [![size](https://img.shields.io/badge/size-30%20MB-brightgreen)](#) |


## Some Notice
```
 The development on port_native is slowed down immensely due to the migration towards seamless GUI EAL interface the port_native will be migrated to a newer architechture which uses GUI frontend that currently under heavy development, thank you for reading this message

```


## Releases

| Build | Update | Release date  | Download |
| ------- | ------ | ------------  | -------- |
| 467~470 | Platform Trial, ssh keyfixes, Content update merge with version 0.06a | July, 2020 | :tickets: [Single Script Archive](https://github.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/releases/tag/Build_46x)
| 450 | Added support for new termux dependencies | June, 2020 | :tickets: [Single Script Archive](https://github.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/releases/tag/Build_450) |
| 416~418 | ServerID implementation, Virtualenv dependencies, etc.. | May, 2020 | :tickets: [Single Script Archive](https://github.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/releases/tag/Build_416) |
| 319~325 | Invading feature, Major bug fixes, Stability improvement, etc.. | April, 2020 | :tickets: [Single Script Archive](https://github.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/releases/tag/Build_319) |
| 260 | Invade neighbors initial for Build 319 | April, 2020 | :tickets: [Single Script Archive](https://github.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/releases/tag/Build_260) |
| 239~257 | Base32 and base64 Encoder, Liveupdate | April, 2020 | [Quality Control failed] |
| 210 | Added Local mode feature & Stability improvements | April, 2020 | :tickets: [Single Script Archive](https://github.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/releases/tag/Build_210) |
| 201 | New Welcome screen & Simpler installation with 'curl' | April, 2020 | :tickets: [Single Script Archive](https://github.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/releases/tag/Build_201) |
| 193 | Custom ssh config & Merge with new main repo | April, 2020 | :tickets: [Single Script Archive](https://github.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/releases/tag/Build_193) |
| 185 | UI Overhaul & Additional bonus decorations | March, 2020 | :tickets: [Single Script Archive](https://github.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/releases/tag/Build_185) |
| 171 | Initial for UI Overhaul | March, 2020 | [Quality Control failed] |
| 129 | Multiplayer Update Merge with version 0.05a   | February, 2020    | :tickets: [Single Script Archive](http://tiny.cc/RaiseTheEmpires_129) |
| 119 | Automatic Host forwarding | January, 2020 | :tickets: [Single Script Archive](http://tiny.cc/EALSERVE119) |
| 99  | Initial Host forwarding | January, 2020 | [Quality Control failed] |
| 57~58 | First initial Release on android x86 and GNU systems natively | January, 2020 | :tickets: [Single Script Archive](http://tiny.cc/RaiseTheEmpires2020_01_24) |
| 20  | Released noArch version | January, 2020 | :tickets: [Single Script Archive](http://tiny.cc/RaiseTheEmpires29TNoArch) |
> Single Script Archive is a single installation file that automates the installation of the server ( Same as a setup wizard in windows but custom built ). Now you don't even need to download the file, just copy paste the command below

## How to install ?
### Installing on Android using Termux
1. Install Termux and Puffin Browser from Play Store
2. Copy paste this command to Termux:
```
curl -s "https://raw.githubusercontent.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/master/empiresservercontainer/devTool/TermuxAutoinstaller" | bash -
./empires-server-termux
bash
```
3. If there's confirmation [Y/n]..?, just type Y
4. Wait until the installation is completed (and make sure your internet is stable)
5. Choose 'StartOnline' to start the server
6. After the empires-server is running, copy the link into the Puffin Browser

*Done!*

If you want to run it next time, just type this:
```
empires-server
```


> Tip: If the installation often failed / takes too long, try to use VPN



### Installing on GNU/Linux/Ubuntu
1. Assuming that you have terminal opened
2. Copy paste this command to terminal
```
sudo curl -s "https://raw.githubusercontent.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/master/empiresservercontainer/devTool/GNULinuxAutoinstaller" | bash -
sudo ./empires-server-gnulinux
bash
```
3. If you want to run it next time, just type this:
```
empires-server
```

### Installing on macOS
1. Open the Terminal app
2. Copy paste these command to terminal
```
/bin/bash -c "$(curl -s "https://raw.githubusercontent.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/master/empiresservercontainer/devTool/macOSautoinstaller" | bash -)"
./empires-server-macOS
```
3. If you want to run it next time, just type this:
```
empires-server
```

### Installing on iOS/iPadOS
1. Install testflight from the appstore
2. Install iSH from this link https://testflight.apple.com/join/97i7KM8O
3. Install photonX for the flash browser
4. Copy paste this command in iSH
```
curl -s "https://raw.githubusercontent.com/Questandachievement7Developer/RaiseTheEmpiresPort_Native/master/empiresservercontainer/devTool/iOSautoinstaller" | bash -
./empires-server-appleiOS
```
7. Allow location usage while using **Do not set it to always allow ( it will killed by the iOS/iPadOS)**
6. If you want to run it next time, just type this:
```
empires-server
```


## Feels confused about the installation setup or anything wanted to implemented in this version?
Feel free to contact to these: <br>
:email: questandachievement@gmail.com  <br>
Personal Discord Yuuta kun#5097 <br>
Our community discord [Discord group](https://discord.gg/xrNE6Hg)
> Don't hestitate to ask we will help you and we won't kill you if you ask so many questions xd


## Reporting Bugs and Contacting
:speech_balloon: Check our Community [Discord group](https://discord.gg/xrNE6Hg)  [![Discord Channel](https://img.shields.io/discord/536575691563466772?label=)](https://discord.gg/xrNE6Hg)



## Tools

:world_map: [Empires & Allies Empire Map Editor](https://github.com/AcidCaos/ea-empire-editor)

:minidisc: [Empires & Allies Servers Support](https://github.com/AcidCaos/raisetheempires/blob/master/SERVERS.md)

## External Links
:beginner: [Empires & Allies Wiki (FANDOM)](https://empiresandallies.fandom.com/wiki/Empires_%26_Allies_Wiki/Main_2)

:clapper: [Empires & Allies Trailer](https://www.youtube.com/watch?v=pAdlyz9miqo)

:smiley_cat: [Empires & Allies subreddit](https://reddit.com/r/EmpiresAndAllies)

## On archives and accessibility to digital heritage
- [Exemption to PCCPSACT](https://www.federalregister.gov/documents/2018/10/26/2018-23241/exemption-to-prohibition-on-circumvention-of-copyright-protection-systems-for-access-control), exemptions to the provision of the Digital Millennium Copyright Act (“DMCA”).
- [EFGAMP](https://efgamp.eu/), the European Federation of Video Game Archives, Museums and Preservation projects.
- [UNESCO PERSIST Programme](https://unescopersist.org/), helps ensure that digital information can continue to be accessed in the future.
- [The Internet Archive](https://archive.org/), a digital library of Internet sites and other cultural artifacts in digital form.
- [BlueMaxima's Flashpoint](https://bluemaxima.org/flashpoint/), the webgame preservation project.

<br><br>
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue)](http://www.gnu.org/licenses/gpl-3.0)
```
Old Empires and Allies preservation project Copyright (C) 2019-2020 by the Raise the Empires team.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
```


empiresservercontainer/changelog.txt
Announcement

Universal Build 46x
---- Port_native Additions and bugfixes
+ Added sshkeyreg() for ssh id_rsa handler for StartOnline game session
+ registrarServerID() has been updated to meet the ssh localhost.run compliant "Adding ssh key"
+ openssl error on previous build caused by rootfs update in termux has been Fixed
+ iOSResumeOp() has been added in order to handle interrupted session unexpectedly and resume whenever its available again
+ Fixes in Starttemplate_iOS
+ Update menu has been disabled for installation breaking concerns
--- Game content Update 0.06a release Nat 11/07/2020 01:29
+ Invade part 4: Neighbor repel, invade, longer sessions, titan Nat 11/07/2020 00:42
+ fixed A Spectrum of Ore quest Willy030125* 30/06/2020 08:46
+ Disable crash log option for servers Nat 20/06/2020 21:50
+ Additional recovered assets (of Willy) Nat 20/06/2020 20:07
+ Servers: Basic documentation AcidCaos 19/06/2020 11:57
+ Profile avatar images Willy030125* 19/06/2020 02:52
+ Servers: login template AcidCaos 18/06/2020 20:50
+ Servers: friends/neighbours page AcidCaos 18/06/2020 19:10
+ fixed self bar heal oshere1111 16/06/2020 11:56
+ fixed secondary abillty Oshere1111 14/06/2020 00:32
+ fix invalid edited saves from breaking the game Nat 13/06/2020 02:34
+ gas poison fix oshere1111 10/06/2020 19:24
+ Flag for current "stable" version questandachievementProjects* 07/06/2020 06:34
+ Chinese(Simplified) fixed index quotation marks AcidCaos 06/06/2020 20:11
+ added Chinese(Simplified) language (unknown revision) AcidCaos 05/06/2020 20:56
+ Invade part 3: invading & repelling Nat 02/06/2020 02:50
+ Element z part 2: fix collection Nat 25/05/2020 17:59
+ Element z part 1 (most fixed except 2 known bugs) Oshere1111 25/05/2020 08:37
+ Element z part 1 (most fixed except 2 known bugs) Oshere1111 25/05/2020 08:33
+ fixed unit buying Oshere1111 25/05/2020 08:27
+ Element z part 1 (most fixed except 2 known bugs) Oshere1111 25/05/2020 00:46
+ deleted dupped trees. releases page AcidCaos 18/05/2020 10:45
+ fixed research crash Oshere1111 08/05/2020 22:51
+ fixed defense shield Oshere1111 04/05/2020 01:42
+ all powerups are shown Oshere1111 01/05/2020 22:28
+ new powerups are usable Oshere1111 01/05/2020 22:12
+ resources gets reduced Oshere1111 29/04/2020 22:47
+ autopickup drops Nat 29/04/2020 22:33
+ fix depleted z-units power-ups out of sync Nat 26/04/2020 00:50
+ obelisk fixed Oshere1111 26/04/2020 00:05
+ xp on hover AcidCaos 24/04/2020 23:08
+ fixed resource needed buying Oshere1111 24/04/2020 15:52
+ resource amount fix Nat 24/04/2020 15:50
+ generic buying item, invade basic battle Nat 24/04/2020 15:40
+ fixed unit buying Oshere1111 24/04/2020 12:03
+ init custom name. favicon fix. AcidCaos 24/04/2020 10:46
+ fixed github repository link Willy030125* 24/04/2020 06:58
+ fixed silence gap in battle music Willy030125* 24/04/2020 06:32
+ fixed resources buying in build menu Oshere1111 23/04/2020 22:39
+ player menu AcidCaos 23/04/2020 12:17
+ adding crewInfo when migrate Oshere1111 22/04/2020 21:07
+ all govt adding crewInfo when migrate Oshere1111 22/04/2020 21:03
+ adding crewInfo when migrate Oshere1111 22/04/2020 15:43
+ crewing buildings Nat 21/04/2020 23:39
+ ability to cancel units, research, produce, ... Nat 21/04/2020 01:05
+ fix quest kill counter Oshere1111 20/04/2020 23:46
+ test2 Oshere1111 20/04/2020 18:05
+ test Oshere1111 20/04/2020 18:01
+ fix build upgrade quests Nat* 20/04/2020 19:35
+ progress fix oshere1111* 19/04/2020 18:50
+ use general consumable oshere1111* 19/04/2020 17:38
+ consumable quest progress oshere1111* 18/04/2020 21:05
+ buy consumables oshere1111* 17/04/2020 01:34
+ progress harvest consumable. powerup counting oshere1111* 16/04/2020 02:10
+ invade neighbors initial Nat 14/04/2020 01:59
+ unlock units (& free unit), LE discounts Nat 05/04/2020 00:47
+ Enabling bonus decorations Nat 23/02/2020 20:51
+ Update README.md questandachievementProjects* 21/02/2020 01:43
+ Public README (II) AcidCaos 20/02/2020 19:27
+ On digital heritage info. AcidCaos 17/02/2020 00:34
+ Added GPL v3 LICENSE (Readme info.) AcidCaos 16/02/2020 23:48
+ Public project. README AcidCaos 15/02/2020 18:44


Universal Build 38x
---- Port_native Addition and bugfixes
+ added warning for android users when trying to start online about their bandwidth usage
+ macOS weird behavior fixes
+ Major compiler overhaul on assets Management and server Patching
+ Rename into more reasonable and logically folder name to assets
+ empires-server Efficiency increased thus performance increases
+ COMPUTERNAME patching implemented
+ Dependencies installed within empiresservercontainer thus become sort of a sandbox using virtualenv
+ Added warning to macOS for using ssh and android for data bandwidth
+ renaming from Mod to staging assets for better understanding
+ serverID regenerate in menu if there is some problem on the connection
+ added update button for easy access to updates
+ Autoupdate has been deprecated in favor of manual update in order for the user to have control over and prevent server breaking
+ Server menu has been updated to be categorized by groups to reduce clutter
+ Some menu has been renamed to get better understanding
+ serverID Fixes login issues
---- Game content updates
+ all powerups are shown
+ new powerups are usable
+ resources gets reduced
+ autopickup drops



The nightly version of universal build 200>
++ Simpler installation command with 'curl'
 + ssh integration to 'curl' command
+ Updated Welcome screen with new logo
+ empires-server now can be start locally to play on PC
+ Compiler fixes
+ Merge with AcidCaos' repo 2020_04_26:
> fix depleted z-units power-ups out of sync Nat 26/04/2020
> obelisk fixed oshere1111 26/04/2020
> xp on hover AcidCaos 25/04/2020
> fixed resource needed buying oshere1111 24/04/2020
> resource amount fix Nat 24/04/2020
> generic buying item, invade basic battle Nat 24/04/2020
> fixed unit buying oshere1111 24/04/2020

empiresservercontainer/compile.sh
#!/bin/bash
#compiler tool
#This tool compiles all of these into one Empires-Server portable executable aarch64
export origindir=$(pwd)
export compileprefix=[EALGNUCompiler]
HEIGHT=15
WIDTH=80
CHOICE_HEIGHT=18
BACKTITLE="empires-server GNU compiler"
TITLE="empires-server GNU compiler"
#installs dependencies python and etc

if [ $(whoami) != 'root' ];then
  echo Launch this script in root
  exit
fi



. ${origindir}/devTool/compileToolset
source ${origindir}/devTool/compileToolset
#install dependencies
dependenciesCompilerInstall

# to make sure that ndk-build or sdkmanager able to call within this script
echo ${PATH}
export PATH=${PATH}:${origindir}/devTool/android_ndk:${origindir}/devTool/android_sdk
echo ${PATH}
#sdkmanager
#ndk-build

#breakpoint

if [ ! -d ${origindir}/rootfs ]; then
  sudo chmod -R 777 ${origindir}
fi
cleanup # cleaning previous interupted build

reposync_rteMain
startupMenu(){
  reset
  cd ${origindir}
if [ -z ${compilealltrigger} ]; then

OPTIONS=(
"" '>>>>> Version <<<<<'
"" "Port_native Build $(cat ${origindir}/buildnumber)"
"" "Compiler Version $(cd ${origindir}/.. ; git log -1 --format=%cd)"
"==========" "===================Compile Action=========================="
"Compile" "Compile packages for distribution"
"buildtest" "check current build stability"
"compileall" "compile all platforms only for termux macOS gnulinux appleiOS(iSH)"
"==========" "===================Repo Action============================="
"repoSync" "sync all repo Its a must!"
"dossucks" "dos2unix everything because DOS format sucks"
"==========" "===================Local Action============================="
"exit" "exit Compiler toolset")



ACTION=$(dialog --clear \
                --backtitle "$BACKTITLE" \
                --title "Select the Action" \
                --menu "$MENU" \
                27 120 $CHOICE_HEIGHT \
                "${OPTIONS[@]}" \
                2>&1 >/dev/tty)

#welp if the devloper is bored and decided to go to this menu instead
if [ -z ${ACTION} ]; then
  startupMenu
  exit
fi
if [ ${ACTION} == "==========" ]; then
startupMenu
exit
fi

if [ ${ACTION} == "exit" ]; then
exit
fi

if [ ${ACTION} == "dossucks" ]; then
stopscrewingupwithmyfilewindows
startupMenu
fi

if [ ${ACTION} == 'repoSync' ]; then
reposync_mainBranch
startupMenu
exit
fi

if [ ${ACTION} == 'localegen' ]; then
xmlTranslator
sleep 5
startupMenu
exit
fi

if [ ${ACTION} == 'buildtest' ]; then
export instTarget="gnulinux"
sanitybuildcheck
startupMenu
exit
fi

if [ ${ACTION} == 'localPushRemote' ]; then
repopush_mainBranch
startupMenu
exit
fi



if [ ${ACTION} == 'Compile' ]; then
compileInitiator
exit
fi

if [ ${ACTION} == "compileall" ]; then
export compilealltrigger=1
listofplatform="termux gnulinux macOS appleiOS"
for a in ${listofplatform}; do
export instTarget="${a}"
bulkBuild
done
exit
fi


fi
}
compileInitiator(){
export instTarget=""
if [ $(uname -m) != "x86_64" ]; then
echo it is recomended to use x86_64 machine to compile this program
fi

if [ -z ${instTarget} ]; then

echo "${compileprefix} Skipping empires-engine x64 Build"
echo "${compileprefix} Launching empires-server arch selection"

OPTIONS=(

"windows" "use wine64 to cross compile to windows"
"termux" "Compiles for Android termux so it can be run on that platform"
"gnulinux" "compiles to standard gnuLinux Format "
"appleiOS" "compiles to iSH iOS runtime (i686) "
"macOS" "compiles to macOS powered by homebrew ")

instTarget=$(dialog --clear \
                --backtitle "$BACKTITLE" \
                --title "Select the target platform" \
                --menu "$MENU" \
                $HEIGHT $WIDTH $CHOICE_HEIGHT \
                "${OPTIONS[@]}" \
                2>&1 >/dev/tty)

fi

if [ -z ${instTarget} ]; then
  startupMenu
  exit
fi

#if [ ${instTarget} != 'termux' ]; then
#Architechture Selection
#                OPTIONS=(
#                "x86_64" "Compiles to x86_64 bin format"
#                "armv8" "compiles to armv8 or aarch64 bin format"
#                "armv7" "compiles to armv7 bin format")
#
#                Arch=$(dialog --clear \
#                                --backtitle "$BACKTITLE" \
#                                --title "Select the target platform" \
#                                --menu "$MENU" \
#                                $HEIGHT $WIDTH $CHOICE_HEIGHT \
#                                "${OPTIONS[@]}" \
#                                2>&1 >/dev/tty)


#else

#fi
#compiles the dependencies and binaries first

#this is really simple Compiler dont make it complicated
buildTHESERVER
startupMenu
}
#_______MAIN_______
startupMenu
